#286 customer invoice mutation doc

andreievg opened the pull request:
Charge to 286

### Changes

* Added Customer Invoice mutation
* Add errors for customer invoice mutation
* Add ability to create anchors without them showing in the menu bar on the right side
* Changed types to have new translations

(slight arrangements in upstream PR for supplier invoice mutation)
(upstream translation branch changes queries for new translations)

I didn't uphold my promise to try to shape BatchReductionBelowZero in a way it can be used to merge into front end state, suggestions welcome (i am not sure what that state is)

joshxg reviewed (changes requested):
Great, thanks for this work! 😀 🎉 

My main comments:
- The differentiation between upsert and insert is confusing. It is made easier to distinguish between the differences if we use update and insert, with upsert being both. I think there is a lot of text that can be tided/removed if the type is a union as the distinction is then obvious.
- Some consistency in the error types would be helpful. I understand these are examples, but it would be good to ensure we can match errors at any level of the hierarchy for the required task. This would mean creating something like (obviously incomplete example):
![image](https://user-images.githubusercontent.com/35858975/133573874-13319bad-cb6e-4a68-b031-31e5e61f7824.png)

> I didn't uphold my promise to try to shape BatchReductionBelowZero in a way it can be used to merge into front end state, suggestions welcome (i am not sure what that state is)

The suggestion wasn't to know the state exactly, but to return errors in the same shape that the mutation was sent. 

For example, as I understand the return object would be:

```
[
    { code: ErrorCode.OtherPartyNameSomething, .. }, 
    { code: ErrorCode.BatchReductionBelowZero, .. }, 
    { code: ErrorCode.BatchReductionBelowZero, .. },
    { code: ErrorCode.BatchReductionBelowZero, .. }
]
```

The suggestion would just be that instead of an array, if the mutation that was sent was:

```
{
    id: "",
    lines: [ {id: "", .. } ]
}
```

Then essentially just adding an `error` property to any invalid object:

```
{
    id: ..
    error: { code: ErrorCode.OtherPartyNameSomething, .. }, 
    lines: [  { id, error: { code: ErrorCode.BatchReductionBelowZero, .. }}, ... ]
}
```

mark-prins reviewed (approved):
:+1: Awesome! thanks again @andreievg 

Had a couple of :nits - mostly though, Josh has beaten me to it and is asking the same sort of things. Main one is why there is the question of distinguishing between insert, update and upsert.

andreievg commented on docs/content/docs/api/mutations.md:
```diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
 +++
 
-## MUTATIONS
+Mutations will happen sequentially if batched, an error will result in cancellation of mutation batch in which it originated and further mutation in the sequence will be discarded (but batched mutation prior to erroneous one would have been executed).
```
> I found this hard to read. Suggest:

updated [upstream](https://github.com/openmsupply/remote-server/pull/334) - f4d9b6b97065d4e841ee727c0afffa5ab91707f0

> I wonder how useful this form of batched mutation request is - it feels like an easy way to create invalid internal consistency. Either the mutations are related, in which case a transaction is better, or they aren't in which case it doesn't matter if one fails.

Agree. I don't know how to disable it (batching is a standard GraphQL feature that async graphql provides)

andreievg commented on docs/content/docs/api/mutations.md:
````diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
+    stockLineId: String
+    numberOfPacks: Number
+}
+```
````
I think this is resolved in our convo about `InputObject` unions

andreievg commented on docs/content/docs/api/mutations.md:
```diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
+    otherPartyId?: string,
+    comment?: string,
+    status?: InvoiceStatus,
+    theirFerefence?: string,
```
i think this is fixed upstream, thanks for noting

andreievg commented on docs/content/docs/api/mutations.md:
```diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
+type InsertCustomerInvoiceLineInput = {
+    itemId: string,
+    stockLineId: string,
+    numberOfPacks: number
```
There is a concept of 'placeholder line' in customer invoices in mSupply, which don't have stockLineId (although stockLineId is compulsory in insert mutation, it's possible that this will be changed, maybe best to keep it ?)

joshxg replied:
Ah of course, thanks 🤦  

This is the same in mSupply for stocktake lines for newly added batches (except mSupply uses a non-null `item_line_ID` for a non-existent item_line). 

Maybe we need to flag this for a better solution as it's not ideal to have trans lines with no item line imo (I will continue to say item line not stock line for approx. 7 months, sorry!).. maybe we need placeholder item lines instead 🙄

andreievg commented on docs/content/docs/api/custom_errors.md:
```diff
@@ -24,7 +24,21 @@ enum CustomErrorCodes {
+  NumberOfPacksMissingInInvoiceLine = "NUMBER_OF_PACKS_MISSING_IN_INVOICE_LINE",
+  BatchReductionBelowZero = "BATCH_REDUCTION_BELOW_ZERO",
+  ItemIdMissingInInvoiceLine = "ITEM_ID_MISSING_IN_INVOICE_LINE",
+  StockLineIdMissingInInvoiceLine =  "STOCK_LINE_ID_MISSING_IN_INVOICE_LINE"
```
Is the concern with bundle size or traffic ? 
* Bundle, how much extra does it add ?
* Traffic, seems like a very infrequent response ?
Either way I think while we deving, human readable might be a good way to go ?

joshxg replied:
Aye yeah sorry not very clear. Bundle size was my concern. I know it's premature. Just, if there are ~500 references to ~10byte error codes in the JS, it could be a couple KB we could save on some bundles (also lazy loading some parts etc, so less).

Maybe something we could optimise in a build step in the future..

andreievg commented on docs/content/docs/api/custom_errors.md:
```diff
@@ -60,4 +75,117 @@ interface RecordNotFoundError extends CustomError {
+interface OtherPartyNotACustomerOfThisStore extends CustomError {
+  code: CustomErrorCodes.OtherPartyNotACustomerOfThisStore,
+  otherPartyId: string,
+}
```
I've update errors to be more hierarchical btw, here (maybe your comments is different now):  updated [upstream](https://github.com/openmsupply/remote-server/pull/334) - f986d654feeab9e04187fb07c2e7fc70e9669c89

I am a bit confused by the comment, types in this document just meant to describe the error (we will be coding them in rust). 

> in general the expressive errors are great for debugging but not ideal for coding.

Not suggesting this is how it would be coded, up to front end, this is just a description of errors (and focus is on the consumer, if this is bad form consumer point of view, let me know exactly what the errors should be and will change, see below)

> I would like to see a more consistent type:

> interface ValidationError<T> extends CustomError {
>  code: ErrorCode,
>  field:  keyof T
> }

Is this different from: https://github.com/openmsupply/remote-server/pull/313#discussion_r704791066

> really like the idea of typing errors if that's possible,

Again i am confused, errors are all typed (quite strongly)

I think I am missing something ?

andreievg commented on docs/content/docs/api/custom_errors.md:
```diff
@@ -24,7 +24,21 @@ enum CustomErrorCodes {
   FirstNotInRange = "FIRST_NOT_IN_RANGE",
   // Singular query
-  RecordNotFound = "RECORD_NOT_FOUND"
+  RecordNotFound = "RECORD_NOT_FOUND",
```
async graphql doesn't know about DB (as per pattern description of error types)

andreievg commented on docs/content/docs/api/mutations.md:
```diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
+    status?: InvoiceStatus,
+    theirFerefence?: string,
+    allInvoiceLines?: InsertCustomerInvoiceLineInput[]
+}
```
it's changed now, idk, my bad

andreievg commented on docs/content/docs/api/mutations.md:
```diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
+    status: InvoiceStatus
+    comment: String
+    theirReference: String
+    allInvoiceLines: [UpsertCustomerInvoiceLineInput]
```
great idea, this was updated in the mutation reshuffle [upstream](https://github.com/openmsupply/remote-server/pull/334) - b4eb01435f9622d3d95be5175e5792d40f50a0ef

andreievg commented on docs/content/docs/api/mutations.md:
````diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
+If **id** is specified, it's deemed an update and the following shape needs to be satisfied
+
+```TypeScript
+type InsertCustomerInvoiceLineInput = {
````
Thanks, it's fixed

andreievg commented on docs/content/docs/api/mutations.md:
````diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
+}
+```
+
+As you can see an insert invoice mutation requires all invoice lines to be of [InsertCustomerInvoiceLineInput](/docs/api/mutations/#customer-invoice-line-insert) type, whereas in update invoice mutation invoice lines can be upserted.
````
Ohh i think this was an error on my part, that line should have read `[InsertCustomerInvoiceLineInput] or [UpdateCustomerInvoice]`

I update it here [upstream](https://github.com/openmsupply/remote-server/pull/334) - b4eb01435f9622d3d95be5175e5792d40f50a0ef

andreievg commented on docs/content/docs/api/mutations.md:
```diff
@@ -12,6 +12,148 @@ template = "docs/page.html"
+
+## CUSTOMER INVOICE LINE
+
+Customer invoice lines are always linked to an invoice, and are mutated via [upsertCustomerInvoice](/docs/api/mutations/#customer-invoice).
```
thanks, yeah. That should be changed now

andreievg commented:
@joshxg @mark-prins 

I've updated to use insert and update on invoice mutations (still lines are upserted for update invoice, but insert only for insert invoice). And changed errors slightly to have some hierarchy. I did the changes in upstream branch https://github.com/openmsupply/remote-server/pull/334, as I've reshuffled things there and there would be too many conflicts if i did it in this branch.

One thing I am still confused about is this comment: https://github.com/openmsupply/remote-server/pull/333#discussion_r709841017

Is it suggesting something to be changed in actual error shapes ? Or is it to do with representing error with typescript generics (if it's the later I have some pushback, it should be very easy to read all possible errors, generics help in dev but I dont' think they would add value in docs ?) I did make this change to be a bit stronger typed with RecordNotFound error: cbf601282bec854e9f0d64b818e7bd7c376d90ff

joshxg commented:
@andreievg Sorry bit confused. Should we just merge and then comment on the supplier invoice one? Are all the changes there?

andreievg commented:
Going to address remaining comments, tomorrow, but want to merge for now: (will merge downstream PR https://github.com/openmsupply/remote-server/pull/335)
https://github.com/openmsupply/remote-server/issues/431
